src/asl.ts

```diff
@@ -2918,6 +2934,285 @@ export class ASL {
+      });
+    };
+    const state = getState();
+    return state ? { node: decl, ...state } : undefined;
+  }
+
+  /**
+   * Generic handler for any type of assignment with either an identifier or a binding pattern (deconstruction).
+   *
+   * Nested BindPatterns are handled recursively.
+   *
+   * const x = 1;
+   * =>
+   * const x = 1;
+   *
+   * const { x } = y;
+   * =>
+   * const x = y.x;
+   *
+   * const { x } = { x: 1 };
+   * =>
+   * const x = 1;
+   *
+   * const { x = 1 } = y;
+   * =>
+   * const x = y.x === undefined ? 1 : y.x;
+   *
+   * const { x = 1 } = {};
+   * =>
+   * const x = 1;
+   *
+   * const { x: { z } } = y;
+   * =>
+   * const z = y.x.z;
+   *
+   * const { x, ...rest } = y;
+   * =>
+   * INVALID - rest is unsupported in objects because ASL doesn't support object manipulation (ex: delete) or field enumeration (ex: keySet).
+   *
+   * const [x] = arr;
+   * =>
+   * const x = arr[0];
+   *
+   * const [,x] = arr;
+   * =>
+   * const x = arr[1];
+   *
+   * const [x,...rest] = arr;
+   * =>
+   * const x = arr[0];
+   * const rest = arr.slice(1);
+   */
+  private evalAssignment(
+    pattern: Identifier | BindingPattern,
+    value: ASLGraph.Output
+  ): ASLGraph.SubState | ASLGraph.NodeState | undefined {
+    // assign an identifier the current value
+    if (isIdentifier(pattern)) {
+      return ASLGraph.passWithInput(
+        {
+          Type: "Pass",
+          ResultPath: `$.${pattern.name}`,
+        },
+        value
+      );
+    } else {
+      const rest = pattern.bindings.find((binding) => binding?.rest);
+      if (rest && isObjectBinding(pattern)) {
+        // TODO: validator
+        throw new SynthError(
+          ErrorCodes.StepFunctions_does_not_support_destructuring_object_with_rest
+        );
+      }
+
+      // run each of the assignments as a sequence of states, they should not rely on each other.
+      const assignments = pattern.bindings.map((binding, i) => {
+        if (!binding) {
+          return undefined;
+        }
+
+        const updatedValue: ASLGraph.Output = (() => {
+          if (isArrayBinding(pattern)) {
+            // const [a] = arr;
+            if (ASLGraph.isJsonPath(value)) {
+              return {
+                jsonPath: `${value.jsonPath}[${i}]`,
+              };
+            } else {
+              // const [a] = [1,2,3];
+              if (!Array.isArray(value.value)) {
+                throw new SynthError(
+                  ErrorCodes.Invalid_Input,
+                  "Expected array binding pattern to be on a reference path or array literal."
+                );
+              }
+              return {
+                value: value.value[i],
+                containsJsonPath: value.containsJsonPath,
+              };
+            }
+          } else {
+            // when `name` is a bindingPattern, propertyName should always be present.
+            const propertyNameExpr = binding.propertyName ?? binding.name;
```

Comment:
I don't understand, can you provide syntax to explain?

What about this case?
```ts
const [{a}]
```

Reply:
`ArrayPattern([BindingElm(name=ObjectPattern([BindingElm(name=Ident("a"))]))])`

But this comment is only relevant to ObjectPattern and not ArrayPattern.

```ts
{a:{b}}
// ObjectPattern([BindingElm(name=ObjectBinding([BindingElm(name=Ident("b"))]),propertyName=Ident("a"))])
{a:b}
// ObjectPattern([BindingElm(name=Ident("b"),propertyName=Ident("a"))])
```

Basically, a binding element either outputs a new name (`name=ident || name=iden && propName=ident`) or chains to a nested binding pattern using the property (`propertName=ident && name=pattern`)